flyteadmin/dataproxy/service.go

```diff
@@ -268,6 +289,21 @@ func (s Service) validateCreateDownloadLinkRequest(req *service.CreateDownloadLi
 	return req, nil
 }
 
+func getExtraHeaders(reference storage.DataReference, contentMd5 string, addMetadata bool) map[string]string {
+	headers := map[string]string{"Content-Length": strconv.Itoa(len(contentMd5)), "Content-MD5": contentMd5}
+	if addMetadata {
+		if strings.HasPrefix(reference.String(), "s3://") {
+			headers[fmt.Sprintf("x-amz-meta-%s", storage.FlyteContentMD5)] = contentMd5
+		} else if strings.HasPrefix(reference.String(), "gs://") {
+			headers[fmt.Sprintf("x-goog-meta-%s", storage.FlyteContentMD5)] = contentMd5
+		} else if strings.HasPrefix(reference.String(), "abfs://") {
+			headers[fmt.Sprintf("x-ms-meta-%s", storage.FlyteContentMD5)] = contentMd5
+			headers["x-ms-blob-type"] = "BlockBlob" // https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob?tabs=microsoft-entra-id#remarks
+		}
```

Comment:
Let's not leak storage layer abstraction into outside layers... no other layers should reason about s3 vs gcs... etc.